sycl/plugins/unified_runtime/ur/adapters/level_zero/program.cpp

```diff
@@ -140,46 +141,55 @@ UR_APIEXPORT ur_result_t UR_APICALL urProgramBuild(
   ZeModuleDesc.pBuildFlags = Options;
   ZeModuleDesc.pConstants = Shim.ze();
 
-  ze_device_handle_t ZeDevice = Context->Devices[0]->ZeDevice;
-  ze_context_handle_t ZeContext = Program->Context->ZeContext;
-  ze_module_handle_t ZeModule = nullptr;
-
   ur_result_t Result = UR_RESULT_SUCCESS;
-  Program->State = ur_program_handle_t_::Exe;
-  ze_result_t ZeResult =
-      ZE_CALL_NOCHECK(zeModuleCreate, (ZeContext, ZeDevice, &ZeModuleDesc,
-                                       &ZeModule, &Program->ZeBuildLog));
-  if (ZeResult != ZE_RESULT_SUCCESS) {
-    // We adjust ur_program below to avoid attempting to release zeModule when
-    // RT calls urProgramRelease().
-    Program->State = ur_program_handle_t_::Invalid;
-    Result = ze2urResult(ZeResult);
-    if (ZeModule) {
-      ZE_CALL_NOCHECK(zeModuleDestroy, (ZeModule));
-      ZeModule = nullptr;
-    }
-  } else {
-    // The call to zeModuleCreate does not report an error if there are
-    // unresolved symbols because it thinks these could be resolved later via a
-    // call to zeModuleDynamicLink.  However, modules created with
-    // urProgramBuild are supposed to be fully linked and ready to use.
-    // Therefore, do an extra check now for unresolved symbols.
-    ZeResult = checkUnresolvedSymbols(ZeModule, &Program->ZeBuildLog);
+  for (auto Device : Context->Devices) {
```

Comment:
This will increase JIT time by a factor of number of devices in the context, which would be seen a perf regressions by some users. Few ideas:

1) Only build on root-devices with the expectation that such build will be suitable for sub-devices too.

2) Lazy build, i.e. do not run real zeModuleCreate until it is not requested on a device where build was not yet performed. But with EAGER_INIT we'd do all builds here.